File: apps/teacher/src/main/java/com/instructure/teacher/adapters/FileUploadCoursesAdapter.java
/*
 * Copyright (C) 2017 - present  Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.instructure.teacher.adapters;

import android.content.Context;
import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.instructure.canvasapi2.models.Course;
import com.instructure.canvasapi2.utils.Logger;
import com.instructure.pandautils.utils.ColorKeeper;
import com.instructure.teacher.R;
import com.instructure.teacher.binders.BaseBinder;

import java.util.ArrayList;
import java.util.List;


public class FileUploadCoursesAdapter extends ArrayAdapter<Course> {

    // Filter Types
    public enum Type{TEACHER, STUDENT, NONE}

    // Member Variables
    private List<Course> mCourses = new ArrayList<>();
    private LayoutInflater mInflater;

    public FileUploadCoursesAdapter(Context context, LayoutInflater inflater, List<Course> courses) {
        super(context, R.layout.canvas_context_spinner_adapter_item, courses);
        mInflater = inflater;
        mCourses = courses;
    }

    public void setCourses(List<Course> courses){
        mCourses = courses;
        notifyDataSetChanged();
    }

    //region Adapter Overrides

    @Override
    public int getCount() {
        return mCourses.size();
    }

    @Override
    public View getDropDownView(int position, View convertView,  ViewGroup parent) {
        final CourseViewHolder viewHolder;

        if(convertView == null) {
            convertView = mInflater.inflate(R.layout.canvas_context_spinner_adapter_item, null);
            viewHolder = new CourseViewHolder();
            viewHolder.title = (TextView)convertView.findViewById(R.id.title);
            viewHolder.indicator = (ImageView) convertView.findViewById(R.id.icon);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (CourseViewHolder)convertView.getTag();
        }

        Course item = mCourses.get(position);

        if(item != null) {
            viewHolder.title.setText(item.getName());
            viewHolder.title.setTypeface(null, Typeface.NORMAL);
            viewHolder.indicator.setVisibility(View.VISIBLE);
            viewHolder.indicator.setBackgroundDrawable(BaseBinder.createIndicatorBackground(
                    ColorKeeper.getOrGenerateColor(item)));
        } else {
            viewHolder.indicator.setVisibility(View.GONE);
            viewHolder.title.setText("");
        }

        return convertView;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final CourseViewHolder viewHolder;

        if(convertView == null) {
            Logger.d("creating view holder, getView");
            convertView = mInflater.inflate(R.layout.canvas_context_spinner_adapter_item, null);
            viewHolder = new CourseViewHolder();
            viewHolder.title = (TextView)convertView.findViewById(R.id.title);
            viewHolder.indicator = (ImageView) convertView.findViewById(R.id.icon);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (CourseViewHolder)convertView.getTag();
        }

        Course item = mCourses.get(position);

        if(item != null) {
            viewHolder.title.setText(item.getName());
            viewHolder.indicator.setVisibility(View.VISIBLE);
            viewHolder.indicator.setBackgroundDrawable(BaseBinder.createIndicatorBackground(
                    ColorKeeper.getOrGenerateColor(item)));
        } else {
            viewHolder.indicator.setVisibility(View.GONE);
            viewHolder.title.setText("");
        }
        return convertView;
    }

    private static class CourseViewHolder {
        TextView title;
        ImageView indicator;
    }

    //endregion

    //region Helpers

    public static List<Course> getFilteredCourseList(List<Course> courses, Type filterType){
        List<Course> newCourseList = new ArrayList<>();

        for(Course course : courses){
            if(shouldAddCourse(course, filterType)){
                newCourseList.add(course);
            }
        }
        return newCourseList;
    }

    private static boolean shouldAddCourse(Course course, Type filterType){
        switch(filterType){
            case STUDENT:
                return course.isStudent();
            case TEACHER:
                return course.isTeacher();
            default:
                return false;
        }
    }

    //endregion
}

File: apps/student/src/main/java/com/emeritus/student/adapter/FileUploadCoursesAdapter.kt
/*
 * Copyright (C) 2016 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.emeritus.student.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.OvalShape
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import com.instructure.canvasapi2.models.Course
import com.instructure.pandautils.utils.backgroundColor
import com.emeritus.student.R

class FileUploadCoursesAdapter(
    context: Context,
    private val inflater: LayoutInflater,
    private var courses: List<Course>
) : ArrayAdapter<Course>(context, R.layout.canvas_context_spinner_adapter_item, courses) {
    enum class Type { TEACHER, STUDENT, NONE }

    fun setCourses(courses: List<Course>) {
        this.courses = courses
        notifyDataSetChanged()
    }

    override fun getCount(): Int = courses.size

    @SuppressLint("InflateParams")
    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        return bindView(courses[position], convertView, false)
    }

    @SuppressLint("InflateParams")
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        return bindView(courses[position], convertView, true)
    }

    @SuppressLint("InflateParams")
    private fun bindView(item: Course, convertView: View?, bold: Boolean): View {
        val viewHolder: CourseViewHolder
        val view: View

        if (convertView == null) {
            view = inflater.inflate(R.layout.canvas_context_spinner_adapter_item, null)
            viewHolder = CourseViewHolder(view.findViewById(R.id.title), view.findViewById(R.id.icon))
            view.tag = viewHolder
        } else {
            view = convertView
            viewHolder = view.tag as CourseViewHolder
        }

        viewHolder.title.text = item.name
        if (!bold) viewHolder.title.setTypeface(null, Typeface.NORMAL)
        viewHolder.indicator.visibility = View.VISIBLE
        viewHolder.indicator.background = ShapeDrawable(OvalShape()).apply { paint.color = item.backgroundColor }

        return view
    }

    private data class CourseViewHolder (val title: TextView, val indicator: ImageView )

    companion object {
        fun getFilteredCourseList(courses: List<Course>, filterType: Type): List<Course> {
            return courses.filter {
                when (filterType) {
                    Type.STUDENT -> it.isStudent
                    Type.TEACHER -> it.isTeacher
                    else -> false
                }
            }
        }
    }

}
